tutorial/lesson_18_parallel_associative_reductions.cpp

```diff
@@ -0,0 +1,325 @@
+        // general) only works for associative reductions. Associative reductions
+        // have the nice property that their results are the same no matter how
+        // the computation is grouped (i.e. split into chunks). If rfactor can't
+        // prove the associativity of a reduction, it will throw an error.
+
+        Buffer<int> halide_result = histogram.realize(8);
+
+        // See figures/lesson_18_hist_rfactor_par.gif for a visualization of
+        // what this did.
+
+        // The equivalent C is:
+        int c_intm[input.height()][8];
+        for (int y = 0; y < input.height(); y++) {
+            for (int x = 0; x < 8; x++) {
+                c_intm[y][x] = 0;
+            }
+        }
+        /* parallel */ for (int y = 0; y < input.height(); y++) {
+            for (int r_x = 0; r_x < input.width(); r_x++) {
+                c_intm[y][input(r_x, y) % 8] += 1;
+            }
+        }
+
+        int c_result[8];
+        for (int x = 0; x < 8; x++) {
+            c_result[x] = 0;
+        }
+        for (int x = 0; x < 8; x++) {
+            for (int r_y = 0; r_y < input.height(); r_y++) {
+                c_result[x] += c_intm[r_y][x];
+            }
+        }
+
+        // Check the answers agree:
+        for (int x = 0; x < 8; x++) {
+            if (c_result[x] != halide_result(x)) {
+                printf("halide_result(%d) = %d instead of %d\n",
+                       x, halide_result(x), c_result[x]);
+                return -1;
+            }
+        }
+    }
+
+    {
+        // Since we can factor associative reductions through the scheduling
+        // directive 'rfactor', it is really easy to explore various factorization
+        // strategies. Given the same serial histogram code:
+        Func histogram("hist_rfactor_vec");
+        histogram(x) = 0;
+        RDom r(0, input.width(), 0, input.height());
+        histogram(input(r.x, r.y) % 8) += 1;
+
+        // Instead of r.y, we call rfactor on r.x this time:
+        Func intermediate = histogram.update().rfactor(r.x, u);
+        // which produces an intermediate that is race-condition free across the
+        // x dimension (denoted by 'u' in the intermediate). This allows us to
+        // vectorize the reduction across the inner dimension:
+        intermediate.compute_root().update().vectorize(u, 8);
+        // Note that since vectorizing the inner dimension changes the order of
+        // computations, this trick only works if the associative reduction is
+        // also commutative. rfactor will ensure these properties hold and will
+        // throw an error if it can't prove those properties.
+
+        // Vectorize the initializations to make things faster.
+        intermediate.vectorize(x, 8);
+        histogram.vectorize(x, 8);
+
+        Buffer<int> halide_result = histogram.realize(8);
+
+        // See figures/lesson_18_hist_rfactor_vec.gif for a visualization of
```

Comment:
Also the flashes are really brief. Would be easier to parse with a longer decay factor